Error "the provider detected that no object exists with the given id" when attempting to import any existing resource

I am attempting to use this tool to import existing resources into Terraform from an Azure environment, but no matter what I do I am getting the following error message:

> While attempting to import an existing object to
> "\<resource\>", the provider detected that no
> object exists with the given id. Only pre-existing objects can be imported;
> check that the id is correct and that it is associated with the provider's
> configured region or endpoint, or use "terraform apply" to create a new
> remote object for this resource.

I have tried the following command:
`aztfexport res --use-azure-cli-cred --log-level TRACE --log-path log.txt --env usgovernment -s xxxxxxxxx /subscriptions/xxxxxxxxxxx/resourceGroups/xxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxx`
but it gives me the above error every time.

I have also tried using the rg import command:
` aztfexport rg --log-level TRACE --log-path log.txt --env usgovernment xxxxxxxxxxxxx`
and it gives me the above error for every single resource in the resource group, including the resource group itself.

I have used the az cli to list the resources successfully. 

All of these resources have public access disabled and are accessible via private endpoint on a vnet. My machine is on a network that can reach resources in the vnet via privatelink.

Below is the (sanitized) log from using the first command:

> 2024-07-16T16:00:57.628-0700 [INFO]  aztfexport: aztfexport starts with config: config.Config{CommonConfig:config.CommonConfig{SubscriptionId:"xxxxxxxxxx", AzureSDKCredential:(*azidentity.AzureCLICredential)(0xc006048080), AzureSDKClientOption:policy.ClientOptions{ClientOptions:policy.ClientOptions{APIVersion:"", Cloud:cloud.Configuration{ActiveDirectoryAuthorityHost:"https://login.microsoftonline.us/", Services:map[cloud.ServiceName]cloud.ServiceConfiguration{"resourceManager":cloud.ServiceConfiguration{Audience:"https://management.core.usgovcloudapi.net", Endpoint:"https://management.usgovcloudapi.net"}}}, Logging:policy.LogOptions{IncludeBody:true, AllowedHeaders:[]string(nil), AllowedQueryParams:[]string(nil)}, Retry:policy.RetryOptions{MaxRetries:0, TryTimeout:0, RetryDelay:0, MaxRetryDelay:0, StatusCodes:[]int(nil), ShouldRetry:(func(*http.Response, error) bool)(nil)}, Telemetry:policy.TelemetryOptions{ApplicationID:"aztfexport(azurerm)", Disabled:false}, TracingProvider:tracing.Provider{newTracerFn:(func(string, string) tracing.Tracer)(nil)}, Transport:exported.Transporter(nil), PerCallPolicies:[]exported.Policy(nil), PerRetryPolicies:[]exported.Policy(nil)}, AuxiliaryTenants:[]string(nil), DisableRPRegistration:false}, OutputDir:"/home/xxxxxxxxxx/Scripts/Synapse_tf_export", OutputFileNames:config.OutputFileNames{TerraformFileName:"", ProviderFileName:"", MainFileName:"", ImportBlockFileName:""}, ProviderVersion:"", DevProvider:false, ProviderName:"azurerm", ContinueOnError:false, BackendType:"local", BackendConfig:[]string(nil), ProviderConfig:map[string]cty.Value(nil), FullConfig:false, Parallelism:10, ModulePath:"", HCLOnly:false, TFClient:tfclient.Client(nil), TelemetryClient:telemetry.AppInsightClient{TelemetryClient:(*appinsights.telemetryClient)(0xc006043a40), subscriptionId:"xxxxxxxxxx", installId:"xxxxxxxxxx", sessionId:"xxxxxxxxxx"}, GenerateImportBlock:false}, ResourceId:"/subscriptions/xxxxxxxxxx/resourceGroups/xxxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxxxxx", ResourceGroupName:"", ARGPredicate:"", MappingFile:"", ResourceNamePattern:"", RecursiveQuery:false, TFResourceName:"res-0", TFResourceType:"", IncludeRoleAssignment:false, IncludeResourceGroup:false}
> 
> 2024-07-16T16:00:57.629-0700 [INFO]  aztfexport: New resource meta
> 2024-07-16T16:00:57.631-0700 [DEBUG] aztfexport: UI update. STATUS: initializing | MSG: tea.WindowSizeMsg{Width:207, Height:39}
> 2024-07-16T16:00:57.631-0700 [DEBUG] aztfexport: UI update. STATUS: initializing | MSG: &meta.MetaResource{baseMeta:meta.baseMeta{subscriptionId:"xxxxxxxxxx", azureSDKCred:(*azidentity.AzureCLICredential)(0xc006048080), azureSDKClientOpt:policy.ClientOptions{ClientOptions:policy.ClientOptions{APIVersion:"", Cloud:cloud.Configuration{ActiveDirectoryAuthorityHost:"https://login.microsoftonline.us/", Services:map[cloud.ServiceName]cloud.ServiceConfiguration{"resourceManager":cloud.ServiceConfiguration{Audience:"https://management.core.usgovcloudapi.net", Endpoint:"https://management.usgovcloudapi.net"}}}, Logging:policy.LogOptions{IncludeBody:true, AllowedHeaders:[]string(nil), AllowedQueryParams:[]string(nil)}, Retry:policy.RetryOptions{MaxRetries:0, TryTimeout:0, RetryDelay:0, MaxRetryDelay:0, StatusCodes:[]int(nil), ShouldRetry:(func(*http.Response, error) bool)(nil)}, Telemetry:policy.TelemetryOptions{ApplicationID:"aztfexport(azurerm)", Disabled:false}, TracingProvider:tracing.Provider{newTracerFn:(func(string, string) tracing.Tracer)(nil)}, Transport:exported.Transporter(nil), PerCallPolicies:[]exported.Policy(nil), PerRetryPolicies:[]exported.Policy(nil)}, AuxiliaryTenants:[]string(nil), DisableRPRegistration:false}, outdir:"/home/xxxxxxxxxx/Scripts/Synapse_tf_export", outputFileNames:config.OutputFileNames{TerraformFileName:"terraform.tf", ProviderFileName:"provider.tf", MainFileName:"main.tf", ImportBlockFileName:"import.tf"}, tf:(*tfexec.Terraform)(nil), resourceClient:(*armresources.Client)(0xc00047e078), providerVersion:"3.99.0", devProvider:false, providerName:"azurerm", backendType:"local", backendConfig:[]string(nil), providerConfig:map[string]cty.Value(nil), fullConfig:false, parallelism:10, generateImportFile:false, hclOnly:false, tfclient:tfclient.Client(nil), moduleAddr:"", moduleDir:"/home/xxxxxxxxxx/Scripts/Synapse_tf_export", importBaseDirs:[]string(nil), importModuleDirs:[]string(nil), importTFs:[]*tfexec.Terraform(nil), originBaseState:[]uint8(nil), baseState:[]uint8(nil), tc:telemetry.AppInsightClient{TelemetryClient:(*appinsights.telemetryClient)(0xc006043a40), subscriptionId:"xxxxxxxxxx", installId:"xxxxxxxxxx", sessionId:"xxxxxxxxxx"}}, AzureId:(*armid.ScopedResourceId)(0xc001d00190), ResourceName:"res-0", ResourceType:""}
> 2024-07-16T16:00:57.635-0700 [INFO]  aztfexport: Init Terraform
> 2024-07-16T16:00:57.812-0700 [INFO]  aztfexport: Find terraform binary at /usr/bin/terraform
> 2024-07-16T16:00:57.812-0700 [INFO]  aztfexport: Init provider
> 2024-07-16T16:00:57.812-0700 [INFO]  aztfexport: Output directory doesn't contain provider setting, create one then
> 2024-07-16T16:00:57.813-0700 [INFO]  aztfexport: Output directory doesn't contain terraform block, create one then
> 2024-07-16T16:00:57.813-0700 [DEBUG] aztfexport: Run "terraform init" for the output directory /home/xxxxxxxxxx/Scripts/Synapse_tf_export
> 2024-07-16T16:01:02.269-0700 [DEBUG] aztfexport: Run "terraform init" for the import directory /tmp/aztfexport-1144619604
> 2024-07-16T16:01:02.269-0700 [DEBUG] aztfexport: Run "terraform init" for the import directory /tmp/aztfexport-1811785705
> 2024-07-16T16:01:02.269-0700 [DEBUG] aztfexport: Run "terraform init" for the import directory /tmp/aztfexport-1404755313
> 2024-07-16T16:01:02.269-0700 [DEBUG] aztfexport: Run "terraform init" for the import directory /tmp/aztfexport-1906308737
> 2024-07-16T16:01:02.269-0700 [DEBUG] aztfexport: Run "terraform init" for the import directory /tmp/aztfexport-350614429
> 2024-07-16T16:01:02.269-0700 [DEBUG] aztfexport: Run "terraform init" for the import directory /tmp/aztfexport-1944491435
> 2024-07-16T16:01:02.269-0700 [DEBUG] aztfexport: Run "terraform init" for the import directory /tmp/aztfexport-3955052224
> 2024-07-16T16:01:02.270-0700 [DEBUG] aztfexport: Run "terraform init" for the import directory /tmp/aztfexport-9591396
> 2024-07-16T16:01:02.270-0700 [DEBUG] aztfexport: Run "terraform init" for the import directory /tmp/aztfexport-1121354430
> 2024-07-16T16:01:02.270-0700 [DEBUG] aztfexport: Run "terraform init" for the import directory /tmp/aztfexport-2830121908
> 2024-07-16T16:01:19.136-0700 [DEBUG] aztfexport: UI update. STATUS: initializing | MSG: aztfexportclient.InitProviderDoneMsg{}
> 2024-07-16T16:01:19.136-0700 [DEBUG] aztfexport: Azure Resource set map to TF resource set
> 2024-07-16T16:01:19.153-0700 [DEBUG] aztfexport: UI update. STATUS: listing Azure resources | MSG: aztfexportclient.ListResourceDoneMsg{List:meta.ImportList{meta.ImportItem{AzureResourceID:(*armid.ScopedResourceId)(0xc001d83db0), TFResourceId:"/subscriptions/xxxxxxxxxx/resourceGroups/xxxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxxxxx", ImportError:error(nil), Imported:false, ValidateError:error(nil), TFAddr:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, TFAddrCache:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, IsRecommended:false, Recommendations:[]string(nil), State:cty.NilVal}}}
> 2024-07-16T16:01:19.155-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: tea.WindowSizeMsg{Width:207, Height:39}
> 2024-07-16T16:01:57.491-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: tea.KeyMsg{Type:-1, Runes:[]int32{119}, Alt:false}
> 2024-07-16T16:01:57.493-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: aztfexportclient.StartImportMsg{List:meta.ImportList{meta.ImportItem{AzureResourceID:(*armid.ScopedResourceId)(0xc001d83db0), TFResourceId:"/subscriptions/xxxxxxxxxx/resourceGroups/xxxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxxxxx", ImportError:error(nil), Imported:false, ValidateError:error(nil), TFAddr:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, TFAddrCache:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, IsRecommended:false, Recommendations:[]string(nil), State:cty.NilVal}}}
> 2024-07-16T16:01:57.493-0700 [DEBUG] aztfexport: UI update. STATUS: importing | MSG: tea.WindowSizeMsg{Width:207, Height:39}
> 2024-07-16T16:01:57.493-0700 [INFO]  aztfexport: Importing /subscriptions/xxxxxxxxxx/resourceGroups/xxxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxxxxx as azurerm_machine_learning_workspace.res-0
> 2024-07-16T16:02:01.566-0700 [ERROR] aztfexport: Importing azurerm_machine_learning_workspace.res-0: exit status 1
> 
> Error: Cannot import non-existent remote object
> 
> While attempting to import an existing object to
> "azurerm_machine_learning_workspace.res-0", the provider detected that no
> object exists with the given id. Only pre-existing objects can be imported;
> check that the id is correct and that it is associated with the provider's
> configured region or endpoint, or use "terraform apply" to create a new
> remote object for this resource.
> 2024-07-16T16:02:01.567-0700 [DEBUG] aztfexport: UI update. STATUS: importing | MSG: aztfexportclient.ImportItemsDoneMsg{Items:[]meta.ImportItem{meta.ImportItem{AzureResourceID:(*armid.ScopedResourceId)(0xc001d83db0), TFResourceId:"/subscriptions/xxxxxxxxxx/resourceGroups/xxxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxxxxx", ImportError:(*fmt.wrapError)(0xc001d04400), Imported:false, ValidateError:error(nil), TFAddr:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, TFAddrCache:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, IsRecommended:false, Recommendations:[]string(nil), State:cty.NilVal}}}
> 2024-07-16T16:02:01.567-0700 [DEBUG] aztfexport: UI update. STATUS: importing | MSG: aztfexportclient.ImportDoneMsg{List:meta.ImportList{meta.ImportItem{AzureResourceID:(*armid.ScopedResourceId)(0xc001d83db0), TFResourceId:"/subscriptions/xxxxxxxxxx/resourceGroups/xxxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxxxxx", ImportError:(*fmt.wrapError)(0xc001d04400), Imported:false, ValidateError:error(nil), TFAddr:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, TFAddrCache:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, IsRecommended:false, Recommendations:[]string(nil), State:cty.NilVal}}}
> 2024-07-16T16:02:01.568-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: tea.WindowSizeMsg{Width:207, Height:39}
> 2024-07-16T16:02:01.584-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: progress.FrameMsg{id:1, tag:1}
> 2024-07-16T16:02:03.731-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: tea.KeyMsg{Type:-1, Runes:[]int32{101}, Alt:false}
> 2024-07-16T16:02:03.732-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: aztfexportclient.ShowImportErrorMsg{Item:meta.ImportItem{AzureResourceID:(*armid.ScopedResourceId)(0xc001d83db0), TFResourceId:"/subscriptions/xxxxxxxxxx/resourceGroups/xxxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxxxxx", ImportError:(*fmt.wrapError)(0xc001d04400), Imported:false, ValidateError:error(nil), TFAddr:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, TFAddrCache:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, IsRecommended:false, Recommendations:[]string(nil), State:cty.NilVal}, Index:0, List:meta.ImportList{meta.ImportItem{AzureResourceID:(*armid.ScopedResourceId)(0xc001d83db0), TFResourceId:"/subscriptions/xxxxxxxxxx/resourceGroups/xxxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxxxxx", ImportError:(*fmt.wrapError)(0xc001d04400), Imported:false, ValidateError:error(nil), TFAddr:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, TFAddrCache:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, IsRecommended:false, Recommendations:[]string(nil), State:cty.NilVal}}}
> 2024-07-16T16:02:04.419-0700 [DEBUG] aztfexport: UI update. STATUS: import error message | MSG: tea.KeyMsg{Type:-1, Runes:[]int32{113}, Alt:false}
> 2024-07-16T16:02:04.420-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: tea.WindowSizeMsg{Width:207, Height:39}
> 2024-07-16T16:02:06.661-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: tea.KeyMsg{Type:-1, Runes:[]int32{101}, Alt:false}
> 2024-07-16T16:02:06.662-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: aztfexportclient.ShowImportErrorMsg{Item:meta.ImportItem{AzureResourceID:(*armid.ScopedResourceId)(0xc001d83db0), TFResourceId:"/subscriptions/xxxxxxxxxx/resourceGroups/xxxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxxxxx", ImportError:(*fmt.wrapError)(0xc001d04400), Imported:false, ValidateError:error(nil), TFAddr:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, TFAddrCache:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, IsRecommended:false, Recommendations:[]string(nil), State:cty.NilVal}, Index:0, List:meta.ImportList{meta.ImportItem{AzureResourceID:(*armid.ScopedResourceId)(0xc001d83db0), TFResourceId:"/subscriptions/xxxxxxxxxx/resourceGroups/xxxxxxxxxx/providers/Microsoft.MachineLearningServices/workspaces/xxxxxxxxxx", ImportError:(*fmt.wrapError)(0xc001d04400), Imported:false, ValidateError:error(nil), TFAddr:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, TFAddrCache:tfaddr.TFAddr{Type:"azurerm_machine_learning_workspace", Name:"res-0"}, IsRecommended:false, Recommendations:[]string(nil), State:cty.NilVal}}}
> 2024-07-16T16:02:06.883-0700 [DEBUG] aztfexport: UI update. STATUS: import error message | MSG: tea.KeyMsg{Type:-1, Runes:[]int32{113}, Alt:false}
> 2024-07-16T16:02:06.884-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: tea.WindowSizeMsg{Width:207, Height:39}
> 2024-07-16T16:02:08.310-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: tea.KeyMsg{Type:-1, Runes:[]int32{113}, Alt:false}
> 2024-07-16T16:02:08.333-0700 [DEBUG] aztfexport: UI update. STATUS: building import list | MSG: aztfexportclient.QuitMsg{}
> 2024-07-16T16:02:08.334-0700 [INFO]  aztfexport: aztfexport ends

--- a/auth_oidc.go
+++ b/auth_oidc.go
@@ -7,7 +7,6 @@ import (
 	"io"
 	"net/http"
 	"net/url"
-	"os"
 
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
 
@@ -21,28 +20,25 @@ type OidcCredential struct {
 	requestToken string
 	requestUrl   string
 
-	token         string
-	tokenFilePath string
+	token string
 
 	cred *azidentity.ClientAssertionCredential
 }
 
 type OidcCredentialOptions struct {
 	azcore.ClientOptions
-	TenantID      string
-	ClientID      string
-	RequestToken  string
-	RequestUrl    string
-	Token         string
-	TokenFilePath string
+	TenantID     string
+	ClientID     string
+	RequestToken string
+	RequestUrl   string
+	Token        string
 }
 
 func NewOidcCredential(options *OidcCredentialOptions) (*OidcCredential, error) {
 	w := &OidcCredential{
-		requestToken:  options.RequestToken,
-		requestUrl:    options.RequestUrl,
-		token:         options.Token,
-		tokenFilePath: options.TokenFilePath,
+		requestToken: options.RequestToken,
+		requestUrl:   options.RequestUrl,
+		token:        options.Token,
 	}
 
 	cred, err := azidentity.NewClientAssertionCredential(options.TenantID, options.ClientID, w.getAssertion, &azidentity.ClientAssertionCredentialOptions{ClientOptions: options.ClientOptions})
@@ -61,15 +57,6 @@ func (w *OidcCredential) GetToken(ctx context.Context, opts policy.TokenRequestO
 func (w *OidcCredential) getAssertion(ctx context.Context) (string, error) {
 	if w.token != "" {
 		return w.token, nil
-	}
-
-	if w.tokenFilePath != "" {
-		idTokenData, err := os.ReadFile(w.tokenFilePath)
-		if err != nil {
-			return "", fmt.Errorf("reading token file: %v", err)
-		}
-
-		return string(idTokenData), nil
 	}
 
 	req, err := http.NewRequestWithContext(ctx, http.MethodGet, w.requestUrl, http.NoBody)

--- a/command_before_func.go
+++ b/command_before_func.go
@@ -50,19 +50,57 @@ func commandBeforeFunc(fset *FlagSet) func(ctx *cli.Context) error {
 				return fmt.Errorf("`--tfclient-plugin-path` must be used together with `--hcl-only`")
 			}
 		}
-		occur := 0
-		for _, ok := range []bool{
-			fset.flagUseEnvironmentCred,
-			fset.flagUseManagedIdentityCred,
-			fset.flagUseAzureCLICred,
-			fset.flagUseOIDCCred,
-		} {
-			if ok {
-				occur += 1
-			}
+
+		if err := conflictArgs([]argDesc{
+			{
+				name:  "--client-id",
+				isSet: fset.flagClientId != "",
+			},
+			{
+				name:  "--client-id-file-path",
+				isSet: fset.flagClientIdFilePath != "",
+			},
+		}); err != nil {
+			return err
 		}
-		if occur > 1 {
-			return fmt.Errorf("only one of `--use-environment-cred`, `--use-managed-identity-cred`, `--use-azure-cli-cred` and `--use-oidc-cred` can be specified")
+
+		if err := conflictArgs([]argDesc{
+			{
+				name:  "--client-certificate",
+				isSet: fset.flagClientCertificate != "",
+			},
+			{
+				name:  "--client-certificate-path",
+				isSet: fset.flagClientCertificatePath != "",
+			},
+		}); err != nil {
+			return err
+		}
+
+		if err := conflictArgs([]argDesc{
+			{
+				name:  "--client-secret",
+				isSet: fset.flagClientSecret != "",
+			},
+			{
+				name:  "--client-secret-file-path",
+				isSet: fset.flagClientSecretFilePath != "",
+			},
+		}); err != nil {
+			return err
+		}
+
+		if err := conflictArgs([]argDesc{
+			{
+				name:  "--oidc-token",
+				isSet: fset.flagOIDCToken != "",
+			},
+			{
+				name:  "--oidc-token-file-path",
+				isSet: fset.flagOIDCTokenFilePath != "",
+			},
+		}); err != nil {
+			return err
 		}
 
 		// Initialize output directory
@@ -186,4 +224,22 @@ The output directory is not empty. Please choose one of actions below:
 		}
 		return nil
 	}
+}
+
+type argDesc struct {
+	name  string
+	isSet bool
+}
+
+func conflictArgs(argDescs []argDesc) error {
+	var conflictArgs []string
+	for _, desc := range argDescs {
+		if desc.isSet {
+			conflictArgs = append(conflictArgs, fmt.Sprintf("%q", desc.name))
+		}
+	}
+	if len(conflictArgs) > 1 {
+		return fmt.Errorf("only one of the followings can be specified: %s", strings.Join(conflictArgs, ", "))
+	}
+	return nil
 }

--- a/command_before_func_test.go
+++ b/command_before_func_test.go
@@ -212,20 +212,6 @@ func TestCommondBeforeFunc(t *testing.T) {
 			},
 			err: "`--hcl-only` only works for local backend",
 		},
-		{
-			name: "--use-environment-cred works",
-			fset: FlagSet{
-				flagUseEnvironmentCred: true,
-			},
-		},
-		{
-			name: "--use-environment-cred with --use-azure-cli failed",
-			fset: FlagSet{
-				flagUseEnvironmentCred: true,
-				flagUseAzureCLICred:    true,
-			},
-			err: "only one of `--use-environment-cred`, `--use-managed-identity-cred`, `--use-azure-cli-cred` and `--use-oidc-cred` can be specified",
-		},
 	}
 
 	for _, tt := range cases {

--- a/cred.go
+++ b/cred.go
@@ -157,7 +157,7 @@ func NewDefaultAzureCredential(logger slog.Logger, opt *DefaultAzureCredentialOp
 
 	chain, err := azidentity.NewChainedTokenCredential(creds, nil)
 	if err != nil {
-		err = multierror.Append(err, fmt.Errorf("Errors collected so far: %v", err))
+		err = multierror.Append(err, fmt.Errorf("Errors from credential build tries: %v", errors))
 		return nil, err
 	}
 	return &DefaultAzureCredential{chain: chain}, nil

--- a/cred.go
+++ b/cred.go
@@ -2,6 +2,8 @@ package main
 
 import (
 	"context"
+	"crypto/x509"
+	"encoding/base64"
 	"fmt"
 	"log/slog"
 
@@ -9,6 +11,7 @@ import (
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
 	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"software.sslmate.com/src/go-pkcs12"
 )
 
 type DefaultAzureCredentialOptions struct {
@@ -57,27 +60,34 @@ func NewDefaultAzureCredential(logger slog.Logger, opt *DefaultAzureCredentialOp
 	}
 
 	logger.Info("Building credential via client certificaite")
-	certs, key, err := azidentity.ParseCertificates([]byte(opt.AuthConfig.ClientCertificate), []byte(opt.AuthConfig.ClientCertificatePassword))
-	if err == nil {
-		if cred, err := azidentity.NewClientCertificateCredential(
-			opt.AuthConfig.TenantID,
-			opt.AuthConfig.ClientID,
-			certs,
-			key,
-			&azidentity.ClientCertificateCredentialOptions{
-				ClientOptions:              opt.ClientOptions,
-				AdditionallyAllowedTenants: opt.AuthConfig.AuxiliaryTenantIDs,
-				DisableInstanceDiscovery:   opt.DisableInstanceDiscovery,
-				SendCertificateChain:       opt.SendCertificateChain,
-			},
-		); err == nil {
-			logger.Info("Successfully built credential via client certificate")
-			creds = append(creds, cred)
-		} else {
-			logger.Warn("Building credential via client certificate failed", "error", err)
-		}
+	if cert, err := base64.StdEncoding.DecodeString(opt.AuthConfig.ClientCertificateEncoded); err != nil {
+		logger.Warn("Building credential via client certificate failed", "error", fmt.Errorf("base64 decoidng certificate: %v", err))
 	} else {
-		logger.Warn("Building credential via client secret failed", "error", fmt.Errorf(`failed to parse certificate: %v`, err))
+		// We are using a 3rd party module for parsing the certificate (the same one as is used by go-azure-sdk/sdk/auth/client_certificate_authorizer.go)
+		// Reason can be found at: https://github.com/Azure/azure-sdk-for-go/issues/22906
+		//certs, key, err := azidentity.ParseCertificates(cert, []byte(opt.AuthConfig.ClientCertificatePassword))
+		key, cert, _, err := pkcs12.DecodeChain(cert, opt.AuthConfig.ClientCertificatePassword)
+		if err == nil {
+			if cred, err := azidentity.NewClientCertificateCredential(
+				opt.AuthConfig.TenantID,
+				opt.AuthConfig.ClientID,
+				[]*x509.Certificate{cert},
+				key,
+				&azidentity.ClientCertificateCredentialOptions{
+					ClientOptions:              opt.ClientOptions,
+					AdditionallyAllowedTenants: opt.AuthConfig.AuxiliaryTenantIDs,
+					DisableInstanceDiscovery:   opt.DisableInstanceDiscovery,
+					SendCertificateChain:       opt.SendCertificateChain,
+				},
+			); err == nil {
+				logger.Info("Successfully built credential via client certificate")
+				creds = append(creds, cred)
+			} else {
+				logger.Warn("Building credential via client certificate failed", "error", err)
+			}
+		} else {
+			logger.Warn("Building credential via client certificate failed", "error", fmt.Errorf(`failed to parse certificate: %v`, err))
+		}
 	}
 
 	if !opt.AuthConfig.UseOIDC {

--- a/cred.go
+++ b/cred.go
@@ -1,0 +1,144 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/Azure/aztfexport/pkg/config"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+)
+
+type DefaultAzureCredentialOptions struct {
+	AuthConfig               config.AuthConfig
+	ClientOptions            azcore.ClientOptions
+	DisableInstanceDiscovery bool
+	// Only applies to certificate credential
+	SendCertificateChain bool
+}
+
+// DefaultAzureCredential is a default credential chain for applications that will deploy to Azure.
+// It attempts to authenticate with each of these credential types, in the following order, stopping
+// when one provides a token:
+//   - [ClientSecretCredential]
+//   - [ClientCertificateCredential]
+//   - [OIDCCredential]
+//   - [ManagedIdentityCredential]
+//   - [AzureCLICredential]
+type DefaultAzureCredential struct {
+	chain *azidentity.ChainedTokenCredential
+}
+
+// NewDefaultAzureCredential creates a DefaultAzureCredential. Pass nil for options to accept defaults.
+func NewDefaultAzureCredential(logger slog.Logger, opt *DefaultAzureCredentialOptions) (*DefaultAzureCredential, error) {
+	var creds []azcore.TokenCredential
+
+	if opt == nil {
+		opt = &DefaultAzureCredentialOptions{}
+	}
+
+	logger.Info("Building credential via client secret")
+	if cred, err := azidentity.NewClientSecretCredential(
+		opt.AuthConfig.TenantID,
+		opt.AuthConfig.ClientID,
+		opt.AuthConfig.ClientSecret,
+		&azidentity.ClientSecretCredentialOptions{
+			ClientOptions:              opt.ClientOptions,
+			AdditionallyAllowedTenants: opt.AuthConfig.AuxiliaryTenantIDs,
+			DisableInstanceDiscovery:   opt.DisableInstanceDiscovery,
+		},
+	); err == nil {
+		logger.Info("Successfully built credential via client secret")
+		creds = append(creds, cred)
+	} else {
+		logger.Warn("Building credential via client secret failed", "error", err)
+	}
+
+	logger.Info("Building credential via client certificaite")
+	certs, key, err := azidentity.ParseCertificates([]byte(opt.AuthConfig.ClientCertificate), []byte(opt.AuthConfig.ClientCertificatePassword))
+	if err == nil {
+		if cred, err := azidentity.NewClientCertificateCredential(
+			opt.AuthConfig.TenantID,
+			opt.AuthConfig.ClientID,
+			certs,
+			key,
+			&azidentity.ClientCertificateCredentialOptions{
+				ClientOptions:              opt.ClientOptions,
+				AdditionallyAllowedTenants: opt.AuthConfig.AuxiliaryTenantIDs,
+				DisableInstanceDiscovery:   opt.DisableInstanceDiscovery,
+				SendCertificateChain:       opt.SendCertificateChain,
+			},
+		); err == nil {
+			logger.Info("Successfully built credential via client certificate")
+			creds = append(creds, cred)
+		} else {
+			logger.Warn("Building credential via client certificate failed", "error", err)
+		}
+	} else {
+		logger.Warn("Building credential via client secret failed", "error", fmt.Errorf(`failed to parse certificate: %v`, err))
+	}
+
+	if !opt.AuthConfig.UseOIDC {
+		logger.Info("OIDC credential skipped")
+	} else {
+		logger.Info("Building credential via OIDC")
+		if cred, err := NewOidcCredential(&OidcCredentialOptions{
+			ClientOptions: opt.ClientOptions,
+			TenantID:      opt.AuthConfig.TenantID,
+			ClientID:      opt.AuthConfig.ClientID,
+			RequestToken:  opt.AuthConfig.OIDCTokenRequestToken,
+			RequestUrl:    opt.AuthConfig.OIDCTokenRequestURL,
+			Token:         opt.AuthConfig.OIDCAssertionToken,
+		}); err == nil {
+			logger.Info("Successfully built credential via OIDC")
+			creds = append(creds, cred)
+		} else {
+			logger.Warn("Building credential via OIDC failed", "error", err)
+		}
+	}
+
+	if !opt.AuthConfig.UseManagedIdentity {
+		logger.Info("managed identity credential skipped")
+	} else {
+		logger.Info("Building credential via managed identity")
+		if cred, err := azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
+			ClientOptions: opt.ClientOptions,
+			ID:            azidentity.ClientID(opt.AuthConfig.ClientID),
+		}); err == nil {
+			logger.Info("Successfully built credential via managed identity")
+			creds = append(creds, cred)
+		} else {
+			logger.Warn("Building credential via managed identity failed", "error", err)
+		}
+	}
+
+	if !opt.AuthConfig.UseAzureCLI {
+		logger.Info("Azure CLI credential skipped")
+	} else {
+		logger.Info("Building credential via Azure CLI")
+		if cred, err := azidentity.NewAzureCLICredential(&azidentity.AzureCLICredentialOptions{
+			AdditionallyAllowedTenants: opt.AuthConfig.AuxiliaryTenantIDs,
+			TenantID:                   opt.AuthConfig.TenantID,
+		}); err == nil {
+			logger.Info("Successfully built credential via Azure CLI")
+			creds = append(creds, cred)
+		} else {
+			logger.Warn("Building credential via Azure CLI failed", "error", err)
+		}
+	}
+
+	chain, err := azidentity.NewChainedTokenCredential(creds, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &DefaultAzureCredential{chain: chain}, nil
+}
+
+// GetToken requests an access token from Azure Active Directory. This method is called automatically by Azure SDK clients.
+func (c *DefaultAzureCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	return c.chain.GetToken(ctx, opts)
+}
+
+var _ azcore.TokenCredential = (*DefaultAzureCredential)(nil)

--- a/cred.go
+++ b/cred.go
@@ -11,6 +11,7 @@ import (
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
 	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/hashicorp/go-multierror"
 	"software.sslmate.com/src/go-pkcs12"
 )
 
@@ -37,6 +38,7 @@ type DefaultAzureCredential struct {
 // NewDefaultAzureCredential creates a DefaultAzureCredential. Pass nil for options to accept defaults.
 func NewDefaultAzureCredential(logger slog.Logger, opt *DefaultAzureCredentialOptions) (*DefaultAzureCredential, error) {
 	var creds []azcore.TokenCredential
+	var errors error
 
 	if opt == nil {
 		opt = &DefaultAzureCredentialOptions{}
@@ -56,18 +58,26 @@ func NewDefaultAzureCredential(logger slog.Logger, opt *DefaultAzureCredentialOp
 		logger.Info("Successfully built credential via client secret")
 		creds = append(creds, cred)
 	} else {
-		logger.Warn("Building credential via client secret failed", "error", err)
+		thisErr := fmt.Errorf("Building credential via client secret failed: %v", err)
+		logger.Warn(thisErr.Error())
+		errors = multierror.Append(errors, thisErr)
 	}
 
 	logger.Info("Building credential via client certificaite")
 	if cert, err := base64.StdEncoding.DecodeString(opt.AuthConfig.ClientCertificateEncoded); err != nil {
-		logger.Warn("Building credential via client certificate failed", "error", fmt.Errorf("base64 decoidng certificate: %v", err))
+		thisErr := fmt.Errorf("Building credential via client certificate failed: base64 decoidng certificate: %v", err)
+		logger.Warn(thisErr.Error())
+		errors = multierror.Append(errors, thisErr)
 	} else {
 		// We are using a 3rd party module for parsing the certificate (the same one as is used by go-azure-sdk/sdk/auth/client_certificate_authorizer.go)
 		// Reason can be found at: https://github.com/Azure/azure-sdk-for-go/issues/22906
 		//certs, key, err := azidentity.ParseCertificates(cert, []byte(opt.AuthConfig.ClientCertificatePassword))
 		key, cert, _, err := pkcs12.DecodeChain(cert, opt.AuthConfig.ClientCertificatePassword)
-		if err == nil {
+		if err != nil {
+			thisErr := fmt.Errorf("Building credential via client certificate failed: failed to parse certificate: %v", err)
+			logger.Warn(thisErr.Error())
+			errors = multierror.Append(errors, thisErr)
+		} else {
 			if cred, err := azidentity.NewClientCertificateCredential(
 				opt.AuthConfig.TenantID,
 				opt.AuthConfig.ClientID,
@@ -79,14 +89,14 @@ func NewDefaultAzureCredential(logger slog.Logger, opt *DefaultAzureCredentialOp
 					DisableInstanceDiscovery:   opt.DisableInstanceDiscovery,
 					SendCertificateChain:       opt.SendCertificateChain,
 				},
-			); err == nil {
+			); err != nil {
+				thisErr := fmt.Errorf("Building credential via client certificate failed: %v", err)
+				logger.Warn(thisErr.Error())
+				errors = multierror.Append(errors, thisErr)
+			} else {
 				logger.Info("Successfully built credential via client certificate")
 				creds = append(creds, cred)
-			} else {
-				logger.Warn("Building credential via client certificate failed", "error", err)
 			}
-		} else {
-			logger.Warn("Building credential via client certificate failed", "error", fmt.Errorf(`failed to parse certificate: %v`, err))
 		}
 	}
 
@@ -101,11 +111,13 @@ func NewDefaultAzureCredential(logger slog.Logger, opt *DefaultAzureCredentialOp
 			RequestToken:  opt.AuthConfig.OIDCTokenRequestToken,
 			RequestUrl:    opt.AuthConfig.OIDCTokenRequestURL,
 			Token:         opt.AuthConfig.OIDCAssertionToken,
-		}); err == nil {
+		}); err != nil {
+			thisErr := fmt.Errorf("Building credential via OIDC failed: %v", err)
+			logger.Warn(thisErr.Error())
+			errors = multierror.Append(errors, thisErr)
+		} else {
 			logger.Info("Successfully built credential via OIDC")
 			creds = append(creds, cred)
-		} else {
-			logger.Warn("Building credential via OIDC failed", "error", err)
 		}
 	}
 
@@ -116,11 +128,13 @@ func NewDefaultAzureCredential(logger slog.Logger, opt *DefaultAzureCredentialOp
 		if cred, err := azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
 			ClientOptions: opt.ClientOptions,
 			ID:            azidentity.ClientID(opt.AuthConfig.ClientID),
-		}); err == nil {
+		}); err != nil {
+			thisErr := fmt.Errorf("Building credential via managed identity failed: %v", err)
+			logger.Warn(thisErr.Error())
+			errors = multierror.Append(errors, thisErr)
+		} else {
 			logger.Info("Successfully built credential via managed identity")
 			creds = append(creds, cred)
-		} else {
-			logger.Warn("Building credential via managed identity failed", "error", err)
 		}
 	}
 
@@ -131,16 +145,19 @@ func NewDefaultAzureCredential(logger slog.Logger, opt *DefaultAzureCredentialOp
 		if cred, err := azidentity.NewAzureCLICredential(&azidentity.AzureCLICredentialOptions{
 			AdditionallyAllowedTenants: opt.AuthConfig.AuxiliaryTenantIDs,
 			TenantID:                   opt.AuthConfig.TenantID,
-		}); err == nil {
+		}); err != nil {
+			thisErr := fmt.Errorf("Building credential via Azure CLI failed: %v", err)
+			logger.Warn(thisErr.Error())
+			errors = multierror.Append(errors, thisErr)
+		} else {
 			logger.Info("Successfully built credential via Azure CLI")
 			creds = append(creds, cred)
-		} else {
-			logger.Warn("Building credential via Azure CLI failed", "error", err)
 		}
 	}
 
 	chain, err := azidentity.NewChainedTokenCredential(creds, nil)
 	if err != nil {
+		err = multierror.Append(err, fmt.Errorf("Errors collected so far: %v", err))
 		return nil, err
 	}
 	return &DefaultAzureCredential{chain: chain}, nil

--- a/flag.go
+++ b/flag.go
@@ -2,6 +2,7 @@ package main
 
 import (
 	"context"
+	"encoding/base64"
 	"fmt"
 	"io"
 	"log/slog"
@@ -276,7 +277,7 @@ func (f FlagSet) buildAuthConfig() (*config.AuthConfig, error) {
 		if err != nil {
 			return nil, fmt.Errorf("reading Client ID from file %q: %v", path, err)
 		}
-		clientId = strings.TrimSpace(string(b))
+		clientId = string(b)
 	}
 
 	clientSecret := f.flagClientSecret
@@ -285,16 +286,16 @@ func (f FlagSet) buildAuthConfig() (*config.AuthConfig, error) {
 		if err != nil {
 			return nil, fmt.Errorf("reading Client secret from file %q: %v", path, err)
 		}
-		clientSecret = strings.TrimSpace(string(b))
+		clientSecret = string(b)
 	}
 
-	clientCert := f.flagClientCertificate
+	clientCertEncoded := f.flagClientCertificate
 	if path := f.flagClientCertificatePath; path != "" {
 		b, err := os.ReadFile(path)
 		if err != nil {
 			return nil, fmt.Errorf("reading Client certificate from file %q: %v", path, err)
 		}
-		clientCert = strings.TrimSpace(string(b))
+		clientCertEncoded = base64.StdEncoding.EncodeToString(b)
 	}
 
 	oidcToken := f.flagOIDCToken
@@ -312,7 +313,7 @@ func (f FlagSet) buildAuthConfig() (*config.AuthConfig, error) {
 		AuxiliaryTenantIDs:        f.flagAuxiliaryTenantIds.Value(),
 		ClientID:                  clientId,
 		ClientSecret:              clientSecret,
-		ClientCertificate:         clientCert,
+		ClientCertificateEncoded:  clientCertEncoded,
 		ClientCertificatePassword: f.flagClientCertificatePassword,
 		OIDCTokenRequestToken:     f.flagOIDCRequestToken,
 		OIDCTokenRequestURL:       f.flagOIDCRequestURL,
@@ -343,7 +344,7 @@ func (f FlagSet) BuildCommonConfig() (config.CommonConfig, error) {
 			return config.CommonConfig{}, fmt.Errorf("creating log file %s: %v", path, err)
 		}
 
-		logger := slog.New(slog.NewTextHandler(f, &slog.HandlerOptions{Level: level}))
+		logger = slog.New(slog.NewTextHandler(f, &slog.HandlerOptions{Level: level}))
 
 		// Enable log for azure sdk
 		os.Setenv("AZURE_SDK_GO_LOGGING", "all") // #nosec G104

--- a/flag.go
+++ b/flag.go
@@ -12,12 +12,10 @@ import (
 	"github.com/Azure/aztfexport/internal/log"
 	"github.com/Azure/aztfexport/pkg/config"
 	"github.com/Azure/aztfexport/pkg/telemetry"
-	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
 	azlog "github.com/Azure/azure-sdk-for-go/sdk/azcore/log"
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
-	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
 	"github.com/gofrs/uuid"
 	"github.com/urfave/cli/v2"
 )
@@ -26,7 +24,6 @@ var flagset FlagSet
 
 type FlagSet struct {
 	// common flags
-	flagEnv                 string
 	flagSubscriptionId      string
 	flagOutputDir           string
 	flagOverwrite           bool
@@ -49,14 +46,23 @@ type FlagSet struct {
 	flagLogLevel            string
 
 	// common flags (auth)
-	flagUseEnvironmentCred     bool
-	flagUseManagedIdentityCred bool
-	flagUseAzureCLICred        bool
-	flagUseOIDCCred            bool
-	flagOIDCRequestToken       string
-	flagOIDCRequestURL         string
-	flagOIDCTokenFilePath      string
-	flagOIDCToken              string
+	flagEnv                       string
+	flagTenantId                  string
+	flagAuxiliaryTenantIds        cli.StringSlice
+	flagClientId                  string
+	flagClientIdFilePath          string
+	flagClientCertificate         string
+	flagClientCertificatePath     string
+	flagClientCertificatePassword string
+	flagClientSecret              string
+	flagClientSecretFilePath      string
+	flagOIDCRequestToken          string
+	flagOIDCRequestURL            string
+	flagOIDCTokenFilePath         string
+	flagOIDCToken                 string
+	flagUseManagedIdentityCred    bool
+	flagUseAzureCLICred           bool
+	flagUseOIDCCred               bool
 
 	// common flags (hidden)
 	hflagMockClient         bool
@@ -100,14 +106,13 @@ func (flag FlagSet) DescribeCLI(mode string) string {
 	args := []string{mode}
 
 	// The following flags are skipped eiter not interesting, or might contain sensitive info:
-	// - flagSubscriptionId
 	// - flagOutputDir
 	// - flagDevProvider
 	// - flagBackendConfig
 	// - all hflags
 
-	if flag.flagEnv != "" {
-		args = append(args, "--env="+flag.flagEnv)
+	if flag.flagSubscriptionId != "" {
+		args = append(args, "--subscription=*")
 	}
 	if flag.flagOverwrite {
 		args = append(args, "--overwrite=true")
@@ -152,17 +157,35 @@ func (flag FlagSet) DescribeCLI(mode string) string {
 		args = append(args, "--generate-import-block=true")
 	}
 
-	if flag.flagUseEnvironmentCred {
-		args = append(args, "--use-environment-cred=true")
+	if flag.flagEnv != "" {
+		args = append(args, "--env="+flag.flagEnv)
 	}
-	if flag.flagUseManagedIdentityCred {
-		args = append(args, "--use-managed-identity-cred=true")
+	if flag.flagTenantId != "" {
+		args = append(args, "--tenant-id=*")
 	}
-	if flag.flagUseAzureCLICred {
-		args = append(args, "--use-azure-cli-cred=true")
+	if v := flag.flagAuxiliaryTenantIds.Value(); len(v) != 0 {
+		args = append(args, fmt.Sprintf("--tenant-id=[%d]", len(v)))
 	}
-	if flag.flagUseOIDCCred {
-		args = append(args, "--use-oidc-cred=true")
+	if flag.flagClientId != "" {
+		args = append(args, "--client-id=*")
+	}
+	if flag.flagClientIdFilePath != "" {
+		args = append(args, "--client-id-file-path="+flag.flagClientIdFilePath)
+	}
+	if flag.flagClientCertificate != "" {
+		args = append(args, "--client-certificate=*")
+	}
+	if flag.flagClientCertificatePath != "" {
+		args = append(args, "--client-certificate-path="+flag.flagClientCertificatePath)
+	}
+	if flag.flagClientCertificatePassword != "" {
+		args = append(args, "--client-certificate-password=*")
+	}
+	if flag.flagClientSecret != "" {
+		args = append(args, "--client-secret=***")
+	}
+	if flag.flagClientSecretFilePath != "" {
+		args = append(args, "--client-secret-file-path="+flag.flagClientSecretFilePath)
 	}
 	if flag.flagOIDCRequestToken != "" {
 		args = append(args, "--oidc-request-token=*")
@@ -175,6 +198,15 @@ func (flag FlagSet) DescribeCLI(mode string) string {
 	}
 	if flag.flagOIDCToken != "" {
 		args = append(args, "--oidc-token=*")
+	}
+	if flag.flagUseManagedIdentityCred {
+		args = append(args, "--use-managed-identity-cred=true")
+	}
+	if flag.flagUseAzureCLICred {
+		args = append(args, "--use-azure-cli-cred=true")
+	}
+	if flag.flagUseOIDCCred {
+		args = append(args, "--use-oidc-cred=true")
 	}
 
 	if flag.hflagTFClientPluginPath != "" {
@@ -237,10 +269,96 @@ func initTelemetryClient(subscriptionId string) telemetry.Client {
 	return telemetry.NewAppInsight(subscriptionId, installId, sessionId)
 }
 
-// buildAzureSDKCredAndClientOpt builds the Azure SDK credential and client option from multiple sources (i.e. environment variables, MSI, Azure CLI).
-func buildAzureSDKCredAndClientOpt(fset FlagSet) (azcore.TokenCredential, *arm.ClientOptions, error) {
+func (f FlagSet) buildAuthConfig() (*config.AuthConfig, error) {
+	clientId := f.flagClientId
+	if path := f.flagClientIdFilePath; path != "" {
+		b, err := os.ReadFile(path)
+		if err != nil {
+			return nil, fmt.Errorf("reading Client ID from file %q: %v", path, err)
+		}
+		clientId = strings.TrimSpace(string(b))
+	}
+
+	clientSecret := f.flagClientSecret
+	if path := f.flagClientSecretFilePath; path != "" {
+		b, err := os.ReadFile(path)
+		if err != nil {
+			return nil, fmt.Errorf("reading Client secret from file %q: %v", path, err)
+		}
+		clientSecret = strings.TrimSpace(string(b))
+	}
+
+	clientCert := f.flagClientCertificate
+	if path := f.flagClientCertificatePath; path != "" {
+		b, err := os.ReadFile(path)
+		if err != nil {
+			return nil, fmt.Errorf("reading Client certificate from file %q: %v", path, err)
+		}
+		clientCert = strings.TrimSpace(string(b))
+	}
+
+	oidcToken := f.flagOIDCToken
+	if path := f.flagOIDCTokenFilePath; path != "" {
+		b, err := os.ReadFile(path)
+		if err != nil {
+			return nil, fmt.Errorf("reading OIDC token from file %q: %v", path, err)
+		}
+		oidcToken = strings.TrimSpace(string(b))
+	}
+
+	c := config.AuthConfig{
+		Environment:               f.flagEnv,
+		TenantID:                  f.flagTenantId,
+		AuxiliaryTenantIDs:        f.flagAuxiliaryTenantIds.Value(),
+		ClientID:                  clientId,
+		ClientSecret:              clientSecret,
+		ClientCertificate:         clientCert,
+		ClientCertificatePassword: f.flagClientCertificatePassword,
+		OIDCTokenRequestToken:     f.flagOIDCRequestToken,
+		OIDCTokenRequestURL:       f.flagOIDCRequestURL,
+		OIDCAssertionToken:        oidcToken,
+		UseAzureCLI:               f.flagUseAzureCLICred,
+		UseManagedIdentity:        f.flagUseManagedIdentityCred,
+		UseOIDC:                   f.flagUseOIDCCred,
+	}
+
+	return &c, nil
+}
+
+// BuildCommonConfig builds the CommonConfig from the FlagSet, except the TFClient, which is built afterwards as it requires a logger.
+func (f FlagSet) BuildCommonConfig() (config.CommonConfig, error) {
+	// Logger is only enabled when the log path is specified.
+	// This is because either interactive/non-interactive mode controls the terminal rendering,
+	// logging to stdout/stderr will impact the rendering.
+	logger := slog.New(slog.NewTextHandler(io.Discard, nil))
+	if path := f.flagLogPath; path != "" {
+		level, err := logLevel(f.flagLogLevel)
+		if err != nil {
+			return config.CommonConfig{}, err
+		}
+
+		// #nosec G304
+		f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0600)
+		if err != nil {
+			return config.CommonConfig{}, fmt.Errorf("creating log file %s: %v", path, err)
+		}
+
+		logger := slog.New(slog.NewTextHandler(f, &slog.HandlerOptions{Level: level}))
+
+		// Enable log for azure sdk
+		os.Setenv("AZURE_SDK_GO_LOGGING", "all") // #nosec G104
+		azlog.SetListener(func(cls azlog.Event, msg string) {
+			logger.Log(context.Background(), log.LevelTrace, msg, "event", cls)
+		})
+	}
+
+	authConfig, err := f.buildAuthConfig()
+	if err != nil {
+		return config.CommonConfig{}, err
+	}
+
 	var cloudCfg cloud.Configuration
-	switch env := fset.flagEnv; strings.ToLower(env) {
+	switch env := f.flagEnv; strings.ToLower(env) {
 	case "public":
 		cloudCfg = cloud.AzurePublic
 	case "usgovernment":
@@ -248,109 +366,40 @@ func buildAzureSDKCredAndClientOpt(fset FlagSet) (azcore.TokenCredential, *arm.C
 	case "china":
 		cloudCfg = cloud.AzureChina
 	default:
-		return nil, nil, fmt.Errorf("unknown environment specified: %q", env)
+		return config.CommonConfig{}, fmt.Errorf("unknown environment specified: %q", env)
 	}
 
-	// Maps the auth related environment variables used in the provider to what azidentity honors
-	if v, ok := os.LookupEnv("ARM_TENANT_ID"); ok {
-		// #nosec G104
-		os.Setenv("AZURE_TENANT_ID", v)
-	}
-	if v, ok := os.LookupEnv("ARM_CLIENT_ID"); ok {
-		// #nosec G104
-		os.Setenv("AZURE_CLIENT_ID", v)
-	}
-	if v, ok := os.LookupEnv("ARM_CLIENT_SECRET"); ok {
-		// #nosec G104
-		os.Setenv("AZURE_CLIENT_SECRET", v)
-	}
-	if v, ok := os.LookupEnv("ARM_CLIENT_CERTIFICATE_PATH"); ok {
-		// #nosec G104
-		os.Setenv("AZURE_CLIENT_CERTIFICATE_PATH", v)
-	}
-
-	clientOpt := &arm.ClientOptions{
+	clientOpt := arm.ClientOptions{
 		ClientOptions: policy.ClientOptions{
 			Cloud: cloudCfg,
 			Telemetry: policy.TelemetryOptions{
-				ApplicationID: fmt.Sprintf("aztfexport(%s)", fset.flagProviderName),
+				ApplicationID: fmt.Sprintf("aztfexport(%s)", f.flagProviderName),
 				Disabled:      false,
 			},
 			Logging: policy.LogOptions{
 				IncludeBody: true,
 			},
 		},
+		AuxiliaryTenants:      authConfig.AuxiliaryTenantIDs,
+		DisableRPRegistration: true,
 	}
 
-	tenantId := os.Getenv("ARM_TENANT_ID")
-	var (
-		cred azcore.TokenCredential
-		err  error
-	)
-	switch {
-	case fset.flagUseEnvironmentCred:
-		cred, err = azidentity.NewEnvironmentCredential(&azidentity.EnvironmentCredentialOptions{
-			ClientOptions: clientOpt.ClientOptions,
-		})
-		if err != nil {
-			return nil, nil, fmt.Errorf("failed to new Environment credential: %v", err)
-		}
-		return cred, clientOpt, nil
-	case fset.flagUseManagedIdentityCred:
-		cred, err = azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
-			ClientOptions: clientOpt.ClientOptions,
-		})
-		if err != nil {
-			return nil, nil, fmt.Errorf("failed to new Managed Identity credential: %v", err)
-		}
-		return cred, clientOpt, nil
-	case fset.flagUseAzureCLICred:
-		cred, err = azidentity.NewAzureCLICredential(&azidentity.AzureCLICredentialOptions{
-			TenantID: tenantId,
-		})
-		if err != nil {
-			return nil, nil, fmt.Errorf("failed to new Azure CLI credential: %v", err)
-		}
-		return cred, clientOpt, nil
-	case fset.flagUseOIDCCred:
-		cred, err = NewOidcCredential(&OidcCredentialOptions{
-			ClientOptions: clientOpt.ClientOptions,
-			TenantID:      tenantId,
-			ClientID:      os.Getenv("ARM_CLIENT_ID"),
-			RequestToken:  fset.flagOIDCRequestToken,
-			RequestUrl:    fset.flagOIDCRequestURL,
-			Token:         fset.flagOIDCToken,
-			TokenFilePath: fset.flagOIDCTokenFilePath,
-		})
-		if err != nil {
-			return nil, nil, fmt.Errorf("failed to new OIDC credential: %v", err)
-		}
-		return cred, clientOpt, nil
-	default:
-		opt := &azidentity.DefaultAzureCredentialOptions{
-			ClientOptions: clientOpt.ClientOptions,
-			TenantID:      tenantId,
-		}
-		cred, err := azidentity.NewDefaultAzureCredential(opt)
-		if err != nil {
-			return nil, nil, fmt.Errorf("failed to new Default credential: %v", err)
-		}
-		return cred, clientOpt, nil
-	}
-}
-
-// BuildCommonConfig builds the CommonConfig from the FlagSet, except the TFClient, which is built afterwards as it requires a logger.
-func (f FlagSet) BuildCommonConfig() (config.CommonConfig, error) {
-	cred, clientOpt, err := buildAzureSDKCredAndClientOpt(f)
+	cred, err := NewDefaultAzureCredential(*logger, &DefaultAzureCredentialOptions{
+		AuthConfig:               *authConfig,
+		ClientOptions:            clientOpt.ClientOptions,
+		DisableInstanceDiscovery: false,
+		SendCertificateChain:     false,
+	})
 	if err != nil {
-		return config.CommonConfig{}, err
+		return config.CommonConfig{}, fmt.Errorf("failed to new credential: %v", err)
 	}
 
 	cfg := config.CommonConfig{
-		Logger:               slog.New(slog.NewTextHandler(io.Discard, nil)),
+		Logger:               logger,
+		AuthConfig:           *authConfig,
 		SubscriptionId:       f.flagSubscriptionId,
 		AzureSDKCredential:   cred,
-		AzureSDKClientOption: *clientOpt,
+		AzureSDKClientOption: clientOpt,
 		OutputDir:            f.flagOutputDir,
 		ProviderVersion:      f.flagProviderVersion,
 		ProviderName:         f.flagProviderName,
@@ -373,32 +422,6 @@ func (f FlagSet) BuildCommonConfig() (config.CommonConfig, error) {
 			MainFileName:        "main.aztfexport.tf",
 			ImportBlockFileName: "import.aztfexport.tf",
 		}
-	}
-
-	// Logger is only enabled when the log path is specified.
-	// This is because either interactive/non-interactive mode controls the terminal rendering,
-	// logging to stdout/stderr will impact the rendering.
-	if path := f.flagLogPath; path != "" {
-		level, err := logLevel(f.flagLogLevel)
-		if err != nil {
-			return config.CommonConfig{}, err
-		}
-
-		// #nosec G304
-		f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0600)
-		if err != nil {
-			return config.CommonConfig{}, fmt.Errorf("creating log file %s: %v", path, err)
-		}
-
-		logger := slog.New(slog.NewTextHandler(f, &slog.HandlerOptions{Level: level}))
-
-		// Enable log for azure sdk
-		os.Setenv("AZURE_SDK_GO_LOGGING", "all") // #nosec G104
-		azlog.SetListener(func(cls azlog.Event, msg string) {
-			logger.Log(context.Background(), log.LevelTrace, msg, "event", cls)
-		})
-
-		cfg.Logger = logger
 	}
 
 	return cfg, nil

--- a/go.mod
+++ b/go.mod
@@ -39,6 +39,7 @@ require (
 	github.com/tidwall/sjson v1.2.5
 	github.com/urfave/cli/v2 v2.24.1
 	github.com/zclconf/go-cty v1.14.1
+	software.sslmate.com/src/go-pkcs12 v0.4.0
 )
 
 require (

--- a/go.mod
+++ b/go.mod
@@ -12,6 +12,7 @@ require (
 	github.com/charmbracelet/lipgloss v0.5.0
 	github.com/gofrs/uuid v3.3.0+incompatible
 	github.com/hashicorp/go-hclog v1.6.3
+	github.com/hashicorp/go-multierror v1.1.1
 	github.com/hashicorp/go-version v1.6.0
 	github.com/hashicorp/hc-install v0.7.0
 	github.com/hashicorp/hcl/v2 v2.17.0
@@ -103,7 +104,6 @@ require (
 	github.com/google/pprof v0.0.0-20211214055906-6f57359322fd // indirect
 	github.com/google/uuid v1.3.1 // indirect
 	github.com/hashicorp/errwrap v1.0.0 // indirect
-	github.com/hashicorp/go-multierror v1.1.1 // indirect
 	github.com/hashicorp/go-plugin v1.4.8 // indirect
 	github.com/hashicorp/hcl v0.0.0-20170504190234-a4b07c25de5f // indirect
 	github.com/hashicorp/terraform-plugin-go v0.14.3 // indirect

--- a/internal/meta/base_meta.go
+++ b/internal/meta/base_meta.go
@@ -207,6 +207,60 @@ func NewBaseMeta(cfg config.CommonConfig) (*baseMeta, error) {
 		}
 	}
 
+	// Merge the auth configs to the provider config, if the config not defined
+	providerConfig := cfg.ProviderConfig
+	if providerConfig == nil {
+		providerConfig = map[string]cty.Value{}
+	}
+
+	setIfNoExist := func(k string, v cty.Value) {
+		if _, ok := providerConfig[k]; !ok {
+			providerConfig[k] = v
+		}
+	}
+	if cfg.SubscriptionId != "" {
+		setIfNoExist("subscription_id", cty.StringVal(cfg.SubscriptionId))
+	}
+	if v := cfg.AuthConfig.Environment; v != "" {
+		setIfNoExist("environment", cty.StringVal(v))
+	}
+	if v := cfg.AuthConfig.TenantID; v != "" {
+		setIfNoExist("tenant_id", cty.StringVal(v))
+	}
+
+	if len(cfg.AuthConfig.AuxiliaryTenantIDs) != 0 {
+		var tenantIds []cty.Value
+		for _, id := range cfg.AuthConfig.AuxiliaryTenantIDs {
+			tenantIds = append(tenantIds, cty.StringVal(id))
+		}
+		setIfNoExist("auxiliary_tenant_ids", cty.ListVal(tenantIds))
+	}
+
+	if v := cfg.AuthConfig.ClientID; v != "" {
+		setIfNoExist("client_id", cty.StringVal(v))
+	}
+	if v := cfg.AuthConfig.ClientSecret; v != "" {
+		setIfNoExist("client_secret", cty.StringVal(v))
+	}
+	if v := cfg.AuthConfig.ClientCertificate; v != "" {
+		setIfNoExist("client_certificate", cty.StringVal(v))
+	}
+	if v := cfg.AuthConfig.ClientCertificatePassword; v != "" {
+		setIfNoExist("client_certificate_password", cty.StringVal(v))
+	}
+	if v := cfg.AuthConfig.OIDCTokenRequestToken; v != "" {
+		setIfNoExist("oidc_request_token", cty.StringVal(v))
+	}
+	if v := cfg.AuthConfig.OIDCTokenRequestURL; v != "" {
+		setIfNoExist("oidc_request_url", cty.StringVal(v))
+	}
+	if v := cfg.AuthConfig.OIDCAssertionToken; v != "" {
+		setIfNoExist("oidc_token", cty.StringVal(v))
+	}
+	setIfNoExist("use_msi", cty.BoolVal(cfg.AuthConfig.UseManagedIdentity))
+	setIfNoExist("use_cli", cty.BoolVal(cfg.AuthConfig.UseAzureCLI))
+	setIfNoExist("use_oidc", cty.BoolVal(cfg.AuthConfig.UseOIDC))
+
 	meta := &baseMeta{
 		logger:             cfg.Logger,
 		subscriptionId:     cfg.SubscriptionId,
@@ -219,7 +273,7 @@ func NewBaseMeta(cfg config.CommonConfig) (*baseMeta, error) {
 		devProvider:        cfg.DevProvider,
 		backendType:        cfg.BackendType,
 		backendConfig:      cfg.BackendConfig,
-		providerConfig:     cfg.ProviderConfig,
+		providerConfig:     providerConfig,
 		providerName:       cfg.ProviderName,
 		fullConfig:         cfg.FullConfig,
 		parallelism:        cfg.Parallelism,
@@ -580,14 +634,15 @@ func (meta *baseMeta) buildTerraformConfig(backendType string) string {
 func (meta *baseMeta) buildProviderConfig() string {
 	f := hclwrite.NewEmptyFile()
 
+	var body *hclwrite.Body
 	if meta.useAzAPI() {
-		f.Body().AppendNewBlock("provider", []string{"azapi"}).Body()
+		body = f.Body().AppendNewBlock("provider", []string{"azapi"}).Body()
 	} else {
-		body := f.Body().AppendNewBlock("provider", []string{"azurerm"}).Body()
+		body = f.Body().AppendNewBlock("provider", []string{"azurerm"}).Body()
 		body.AppendNewBlock("features", nil)
-		for k, v := range meta.providerConfig {
-			body.SetAttributeValue(k, v)
-		}
+	}
+	for k, v := range meta.providerConfig {
+		body.SetAttributeValue(k, v)
 	}
 	return string(f.Bytes())
 }

--- a/internal/meta/base_meta.go
+++ b/internal/meta/base_meta.go
@@ -173,11 +173,8 @@ func NewBaseMeta(cfg config.CommonConfig) (*baseMeta, error) {
 	// #nosec G104
 	os.Setenv("AZURE_HTTP_USER_AGENT", cfg.AzureSDKClientOption.Telemetry.ApplicationID)
 
-	// Avoid the AzureRM provider to call the expensive RP listing API, repeatedly.
-	// #nosec G104
+	// Disable AzureRM provider's enahnced validation, which will cause RP listing, that is expensive.
 	os.Setenv("ARM_PROVIDER_ENHANCED_VALIDATION", "false")
-	// #nosec G104
-	os.Setenv("ARM_SKIP_PROVIDER_REGISTRATION", "true")
 
 	outputFileNames := cfg.OutputFileNames
 	if outputFileNames.TerraformFileName == "" {
@@ -207,7 +204,7 @@ func NewBaseMeta(cfg config.CommonConfig) (*baseMeta, error) {
 		}
 	}
 
-	// Merge the auth configs to the provider config, if the config not defined
+	// Update provider config if not explicitly defined
 	providerConfig := cfg.ProviderConfig
 	if providerConfig == nil {
 		providerConfig = map[string]cty.Value{}
@@ -218,6 +215,8 @@ func NewBaseMeta(cfg config.CommonConfig) (*baseMeta, error) {
 			providerConfig[k] = v
 		}
 	}
+
+	// Update provider config for auth config
 	if cfg.SubscriptionId != "" {
 		setIfNoExist("subscription_id", cty.StringVal(cfg.SubscriptionId))
 	}
@@ -242,7 +241,7 @@ func NewBaseMeta(cfg config.CommonConfig) (*baseMeta, error) {
 	if v := cfg.AuthConfig.ClientSecret; v != "" {
 		setIfNoExist("client_secret", cty.StringVal(v))
 	}
-	if v := cfg.AuthConfig.ClientCertificate; v != "" {
+	if v := cfg.AuthConfig.ClientCertificateEncoded; v != "" {
 		setIfNoExist("client_certificate", cty.StringVal(v))
 	}
 	if v := cfg.AuthConfig.ClientCertificatePassword; v != "" {
@@ -260,6 +259,9 @@ func NewBaseMeta(cfg config.CommonConfig) (*baseMeta, error) {
 	setIfNoExist("use_msi", cty.BoolVal(cfg.AuthConfig.UseManagedIdentity))
 	setIfNoExist("use_cli", cty.BoolVal(cfg.AuthConfig.UseAzureCLI))
 	setIfNoExist("use_oidc", cty.BoolVal(cfg.AuthConfig.UseOIDC))
+
+	// Update provider config for provider registration
+	setIfNoExist("skip_provider_registration", cty.BoolVal(true))
 
 	meta := &baseMeta{
 		logger:             cfg.Logger,

--- a/main.go
+++ b/main.go
@@ -262,13 +262,13 @@ func main() {
 		&cli.StringFlag{
 			Name:        "client-certificate",
 			EnvVars:     []string{"AZTFEXPORT_CLIENT_CERTIFICATE", "ARM_CLIENT_CERTIFICATE"},
-			Usage:       "The Azure client certificate",
+			Usage:       "A base64-encoded PKCS#12 bundle to be used as the client certificate for authentication",
 			Destination: &flagset.flagClientCertificate,
 		},
 		&cli.StringFlag{
 			Name:        "client-certificate-path",
 			EnvVars:     []string{"AZTFEXPORT_CLIENT_CERTIFICATE_PATH", "ARM_CLIENT_CERTIFICATE_PATH"},
-			Usage:       "The Azure client certificate file path",
+			Usage:       "The path to the Client Certificate (PKCS#12) associated with the Service Principal which should be used",
 			Destination: &flagset.flagClientCertificatePath,
 		},
 		&cli.StringFlag{

--- a/main.go
+++ b/main.go
@@ -93,16 +93,7 @@ func prepareConfigFile(ctx *cli.Context) error {
 func main() {
 	commonFlags := []cli.Flag{
 		&cli.StringFlag{
-			Name: "env",
-			// Honor the "ARM_ENVIRONMENT" as is used by the provider, for easier use.
-			EnvVars:     []string{"AZTFEXPORT_ENV", "ARM_ENVIRONMENT"},
-			Usage:       `The cloud environment, can be one of "public", "usgovernment" and "china"`,
-			Destination: &flagset.flagEnv,
-			Value:       "public",
-		},
-		&cli.StringFlag{
-			Name: "subscription-id",
-			// Honor the "ARM_SUBSCRIPTION_ID" as is used by the provider, for easier use.
+			Name:        "subscription-id",
 			EnvVars:     []string{"AZTFEXPORT_SUBSCRIPTION_ID", "ARM_SUBSCRIPTION_ID"},
 			Aliases:     []string{"s"},
 			Usage:       "The subscription id",
@@ -237,29 +228,66 @@ func main() {
 		},
 
 		// Common flags (auth)
-		&cli.BoolFlag{
-			Name:        "use-environment-cred",
-			EnvVars:     []string{"AZTFEXPORT_USE_ENVIRONMENT_CRED"},
-			Usage:       "Explicitly use the environment variables to do authentication",
-			Destination: &flagset.flagUseEnvironmentCred,
+		&cli.StringFlag{
+			Name:        "env",
+			EnvVars:     []string{"AZTFEXPORT_ENV", "ARM_ENVIRONMENT"},
+			Usage:       `The cloud environment, can be one of "public", "usgovernment" and "china"`,
+			Destination: &flagset.flagEnv,
+			Value:       "public",
 		},
-		&cli.BoolFlag{
-			Name:        "use-managed-identity-cred",
-			EnvVars:     []string{"AZTFEXPORT_USE_MANAGED_IDENTITY_CRED"},
-			Usage:       "Explicitly use the managed identity that is provided by the Azure host to do authentication",
-			Destination: &flagset.flagUseManagedIdentityCred,
+		&cli.StringFlag{
+			Name:        "tenant-id",
+			EnvVars:     []string{"AZTFEXPORT_TENANT_ID", "ARM_TENANT_ID"},
+			Usage:       "The Azure tenant id",
+			Destination: &flagset.flagTenantId,
 		},
-		&cli.BoolFlag{
-			Name:        "use-azure-cli-cred",
-			EnvVars:     []string{"AZTFEXPORT_USE_AZURE_CLI_CRED"},
-			Usage:       "Explicitly use the Azure CLI to do authentication",
-			Destination: &flagset.flagUseAzureCLICred,
+		&cli.StringSliceFlag{
+			Name:        "auxiliary-tenant-ids",
+			EnvVars:     []string{"AZTFEXPORT_AUXILIARY_TENANT_IDS", "ARM_AUXILIARY_TENANT_IDS"},
+			Usage:       "The Azure auxiliary tenant ids (comma separated)",
+			Destination: &flagset.flagAuxiliaryTenantIds,
 		},
-		&cli.BoolFlag{
-			Name:        "use-oidc-cred",
-			EnvVars:     []string{"AZTFEXPORT_USE_OIDC_CRED"},
-			Usage:       "Explicitly use the OIDC to do authentication",
-			Destination: &flagset.flagUseOIDCCred,
+		&cli.StringFlag{
+			Name:        "client-id",
+			EnvVars:     []string{"AZTFEXPORT_CLIENT_ID", "ARM_CLIENT_ID"},
+			Usage:       "The Azure client id",
+			Destination: &flagset.flagClientId,
+		},
+		&cli.StringFlag{
+			Name:        "client-id-file-path",
+			EnvVars:     []string{"AZTFEXPORT_CLIENT_ID_FILE_PATH", "ARM_CLIENT_ID_FILE_PATH"},
+			Usage:       "The Azure client id file path",
+			Destination: &flagset.flagClientIdFilePath,
+		},
+		&cli.StringFlag{
+			Name:        "client-certificate",
+			EnvVars:     []string{"AZTFEXPORT_CLIENT_CERTIFICATE", "ARM_CLIENT_CERTIFICATE"},
+			Usage:       "The Azure client certificate",
+			Destination: &flagset.flagClientCertificate,
+		},
+		&cli.StringFlag{
+			Name:        "client-certificate-path",
+			EnvVars:     []string{"AZTFEXPORT_CLIENT_CERTIFICATE_PATH", "ARM_CLIENT_CERTIFICATE_PATH"},
+			Usage:       "The Azure client certificate file path",
+			Destination: &flagset.flagClientCertificatePath,
+		},
+		&cli.StringFlag{
+			Name:        "client-certificate-password",
+			EnvVars:     []string{"AZTFEXPORT_CLIENT_CERTIFICATE_PASSWORD", "ARM_CLIENT_CERTIFICATE_PASSWORD"},
+			Usage:       "The password associated with the Client Certificate",
+			Destination: &flagset.flagClientCertificatePassword,
+		},
+		&cli.StringFlag{
+			Name:        "client-secret",
+			EnvVars:     []string{"AZTFEXPORT_CLIENT_SECRET", "ARM_CLIENT_SECRET"},
+			Usage:       "The Azure client secret",
+			Destination: &flagset.flagClientSecret,
+		},
+		&cli.StringFlag{
+			Name:        "client-secret-file-path",
+			EnvVars:     []string{"AZTFEXPORT_CLIENT_SECRET_FILE_PATH", "ARM_CLIENT_SECRET_FILE_PATH"},
+			Usage:       "The Azure client secret file path",
+			Destination: &flagset.flagClientSecretFilePath,
 		},
 		&cli.StringFlag{
 			Name:        "oidc-request-token",
@@ -284,6 +312,27 @@ func main() {
 			EnvVars:     []string{"AZTFEXPORT_OIDC_TOKEN", "ARM_OIDC_TOKEN"},
 			Usage:       "The ID token when authenticating using OIDC",
 			Destination: &flagset.flagOIDCToken,
+		},
+		&cli.BoolFlag{
+			Name:        "use-managed-identity-cred",
+			EnvVars:     []string{"AZTFEXPORT_USE_MANAGED_IDENTITY_CRED", "ARM_USE_MSI"},
+			Usage:       "Explicitly use the managed identity that is provided by the Azure host to do authentication",
+			Destination: &flagset.flagUseManagedIdentityCred,
+			Value:       false,
+		},
+		&cli.BoolFlag{
+			Name:        "use-azure-cli-cred",
+			EnvVars:     []string{"AZTFEXPORT_USE_AZURE_CLI_CRED", "ARM_USE_CLI"},
+			Usage:       "Explicitly use the Azure CLI to do authentication",
+			Destination: &flagset.flagUseAzureCLICred,
+			Value:       true,
+		},
+		&cli.BoolFlag{
+			Name:        "use-oidc-cred",
+			EnvVars:     []string{"AZTFEXPORT_USE_OIDC_CRED", "ARM_USE_OIDC"},
+			Usage:       "Explicitly use the OIDC to do authentication",
+			Destination: &flagset.flagUseOIDCCred,
+			Value:       false,
 		},
 
 		// Hidden flags

--- a/pkg/config/auth_config.go
+++ b/pkg/config/auth_config.go
@@ -1,0 +1,20 @@
+package config
+
+type AuthConfig struct {
+	Environment        string
+	TenantID           string
+	AuxiliaryTenantIDs []string
+
+	ClientID                  string
+	ClientSecret              string
+	ClientCertificate         string
+	ClientCertificatePassword string
+
+	OIDCTokenRequestToken string
+	OIDCTokenRequestURL   string
+	OIDCAssertionToken    string
+
+	UseAzureCLI        bool
+	UseManagedIdentity bool
+	UseOIDC            bool
+}

--- a/pkg/config/auth_config.go
+++ b/pkg/config/auth_config.go
@@ -7,7 +7,7 @@ type AuthConfig struct {
 
 	ClientID                  string
 	ClientSecret              string
-	ClientCertificate         string
+	ClientCertificateEncoded  string
 	ClientCertificatePassword string
 
 	OIDCTokenRequestToken string

--- a/pkg/config/config.go
+++ b/pkg/config/config.go
@@ -42,6 +42,9 @@ type OutputFileNames struct {
 
 type CommonConfig struct {
 	Logger *slog.Logger
+	// AuthConfig specifies the authentication config for provider
+	// This doesn't impact the auth of the tool itself (e.g. listing resources), which is controlled by AzureSDKCredential field.
+	AuthConfig AuthConfig
 	// SubscriptionId specifies the user's Azure subscription id.
 	SubscriptionId string
 	// AzureSDKCredential specifies the Azure SDK token credential

